joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -630,6 +609,47 @@ void JointTrajectoryController::feedback_setup_callback(
     std::bind(&RealtimeGoalHandle::runNonRealtime, rt_active_goal_));
 }
 
+void JointTrajectoryController::fill_partial_goal(
+  std::shared_ptr<trajectory_msgs::msg::JointTrajectory> trajectory_msg) const
+{
+  // joint names in the goal are a subset of existing joints, as checked in goal_callback
+  // so if the size matches, the goal contains all controller joints
+  if (joint_names_.size() == trajectory_msg->joint_names.size()) {
+    return;
+  }
+
+  trajectory_msg->joint_names.reserve(joint_names_.size());
+
+  for (auto index = 0ul; index < joint_names_.size(); ++index) {
+    {
+      if (std::find(
+          trajectory_msg->joint_names.begin(), trajectory_msg->joint_names.end(),
+          joint_names_[index]) != trajectory_msg->joint_names.end())
+      {
+        // joint found on msg
+        continue;
+      }
+      trajectory_msg->joint_names.push_back(joint_names_[index]);
+
+      const auto & joint_state = registered_joint_state_handles_[index];
+      for (auto & it : trajectory_msg->points) {
+        // Assume hold position with 0 velocity and acceleration for missing joints
+        it.positions.push_back(joint_state->get_position());
```

Comment:
This function is called on a separate thread from the main JTC updating loop, you'll get some kind of race condition if you call `joint_state->get_position()` here. 

`traj_external_point_ptr_->update(msg);` is where we sync up, we could put some kind of coordination or delay it until the first parsing of the message (this is how I did immediate paybacks of trajectory messages where messages are sent with time 0). Perhaps we could use a boolean per joint to indicate whether to retain the last known joint position.

Reply:
Thanks for noticing this, it is indeed a problem.

Furthermore, On the [topic callback](https://github.com/ros-controls/ros2_controllers/pull/68/files/c910ecc50cb3c53bd6a6c88690f701660de85b43#diff-fbf197065df5790eaa93ff95fc223e8bR332) we update the trajectory without acquiring the `trajectory_mtx_` which is another race condition.

Probably when we receive a trajectory (whether from topic or action) on an "auxiliary thread" we should do some basic checks to ensure that we can accept it, and store it somewhere protected by a mutex.

Then on the next `update` call on the "main thread", we should process it, fill the partial goals and do whatever needs to be done.

Something along these lines I believe would fix these issues.

Reply:
I've pushed a fix for these issues on this same branch. https://github.com/ros-controls/ros2_controllers/pull/68/commits/188c0e405d8d0db0cdd6c873e41bf9a04d4ffb06